File: benchmarks/Ruby/storage.rb
require_relative 'som'

class Storage < Benchmark

  def initialize
    @count = 0
  end

  def benchmark
    random = Random.new
    @count = 0
    build_tree_depth(7, random)
    @count
  end

  def verify_result(result)
    5461 == result
  end

  def build_tree_depth(depth, random)
    @count += 1

    if depth == 1
      Array.new(random.next % 10 + 1)
    else
      Array.new(4) { build_tree_depth(depth - 1, random) }
    end
  end
end

File: benchmarks/Rust/src/storage.rs
use crate::benchmark::Benchmark;
use crate::som::random::Random;
use std::any::Any;

#[derive(Default)]
pub struct Storage {
    count: i32,
}

impl Benchmark for Storage {
    fn benchmark(&self) -> Box<dyn Any> {
        let mut random = Random::default();
        let mut run = Self::default();
        run.build_tree_depth(7, &mut random);
        Box::new(run.count)
    }

    fn verify_result(&self, result: Box<dyn Any>) -> bool {
        if let Ok(count) = result.downcast::<i32>() {
            5461 == *count
        } else {
            false
        }
    }
}

impl Storage {
    fn build_tree_depth(&mut self, depth: i32, random: &mut Random) -> Box<dyn Any> {
        self.count += 1;
        if depth == 1 {
            Box::new([random.next() % 10 + 1])
        } else {
            let mut build = || self.build_tree_depth(depth - 1, random);
            Box::new([build(), build(), build(), build()])
        }
    }
}
